text/0074-stateful-multisig-pallet.md

```diff
@@ -0,0 +1,585 @@
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 15 February 2024                                                                            |
+| **Description** | Add Enhanced Multisig Pallet to System chains                                               |
+| **Authors**     | Abdelrahman Soliman   (Boda)                                                                |
+
+## Summary
+
+A pallet to facilitate enhanced multisig accounts. The main enhancement is that we store a multisig account in the state with related info (signers, threshold,..etc). The module affords enhanced control over administrative operations such as adding/removing signers, changing the threshold, account deletion, canceling an existing proposal. Each signer can approve/reject a proposal while still exists. The proposal is **not** intended for migrating or getting rid of existing multisig. It's to allow both options to coexist.
+
+For the rest of the RFC We use the following terms:
+
+* `proposal` to refer to an extrinsic that is to be dispatched from a multisig account after getting enough approvals.
+* `Stateful Multisig` to refer to the proposed pallet.
+* `Stateless Multisig` to refer to the current multisig pallet in polkadot-sdk.
+
+## Motivation
+
+### Problem
+
+Entities in the Polkadot ecosystem need to have a way to manage their funds and other operations in a secure and efficient way. Multisig accounts are a common way to achieve this. Entities by definition change over time, members of the entity may change, threshold requirements may change, and the multisig account may need to be deleted. For even more enhanced hierarchical control, the multisig account may need to be controlled by other multisig accounts.
+
+Current native solutions for multisig operations are less optimal, performance-wise (as we'll explain later in the RFC), and lack fine-grained control over the multisig account.
+
+#### Stateless Multisig
+
+We refer to current [multisig pallet in polkadot-sdk](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame/multisig) because the multisig account is only derived and not stored in the state. Although deriving the account is determinsitc as it relies on exact users (sorted) and thershold to derive it. This does not allow for control over the multisig account. It's also tightly coupled to exact users and threshold. This makes it hard for an organization to manage existing accounts and to change the threshold or add/remove signers.
+
+We believe as well that the stateless multisig is not efficient in terms of block footprint as we'll show in the performance section.
+
+#### Pure Proxy
+
+Pure proxy can achieve having a stored and determinstic multisig account from different users but it's unneeded complexity as a way around the limitations of the current multisig pallet. It doesn't also have the same fine grained control over the multisig account.
```

Comment:
As long as the complexity of pure proxy is not leaked to the end user (via a frontend), it should be fine. Stateless implementation should be simpler I am assuming as a counter argument.

Reply:
Using pure-proxy still needs stateless multisig pallet as far as I understand. That means even worse block footprint than the quadratic multisig already. [reference](https://www.youtube.com/watch?v=iGRoGstB_pQ&t=92s)

Reply:
@Tbaut may have more insights on this as well. WDYT Tbaut?

Reply:
What's the question exactly? I don't see the need of pure proxy usage if this pallet is used.

Reply:
Ankan was asking why use this pallet if we already have pure proxy as a counter argument.

Reply:
There are many answers to this.
- pure proxies aren't (yet) a thing cross chain
- the end user complexity is much much higher with pure proxies, also for new users smart contract multisig are widely known while pure proxies are obscure. 
- you can shoot yourself in the foot by deleting the proxy, and effectively loosing access to funds with pure proxies. 
- the problem that I mentioned in my other comment is fundamental. If ppl can't easily verify, they won't. Hence, having the ability to host the call is paramount. Otherwise you rely on centralized 3rd party, it's not related to pure proxies, but it's a very important addition if this pallet supports it.

Reply:
Thanks @tbaut! I added these points to the RFC under your name as well